### PR TITLE
[Hack] WASD movement

### PR DESCRIPTION
**Pull request by afkvido**
### This pull request brings WASD movement into Prodigy Math.

- [X] Only Phasing **or** movement can be on at the same time (or none)
- [X] WASD Phasing hack
   - [X] Manually edit the player's location.
   - [ ] Make WASD movement appear **smooth**.
- [X] WASD Movement hack
   - [ ] Hijack the mouse
   - [ ] Click (at the right position), see if that does anything
   - [ ] If it did nothing, then you ran into a wall.
- [ ] Hack is on by default
    - [ ] Decide which one
- [ ] Support for Harmony Island and Crystal Caverns

### CODE REVIEW

**Comment by afkvido** [comment]
Not ready quite yet, still needs to be a bit smoother.

**Comment by jat354155** [comment]
How to make collision, divide the movement by (speed)/10, if it touches a wall, go back by (speed)/-7.5 . Using this method has worked for me a lot, also prodigy devs do be slacking tho 
image

**Comment by afkvido** [comment]
But why would I want to make colliosons? I do like the phasing, I've found it quite useful.

**Comment by Erisfiregamer1** [comment]
For people who don't want to- well, phase through walls.
Just make it a togle, "Disable Collisions"

**Review by Erisfiregamer1 — APPROVED** [review]
We *could* release this hack now with no repercussions.
But let's finish making it smooth and add collisions first. Then I'll hit Merge.

**Comment by afkvido** [comment]
Bet, let's polish this.
Anyways, any idea on how we're gonna add wall checks?

**Comment by Erisfiregamer1** [comment]
The same Prodigy does it. Just find whatever they check to ensure players don't noclip through walls and use that,.

**Comment by jat354155** [comment]
Ok better description of the script I described,

**Repeat 10
|change (x/y) by speed/10
|if touching wall then
||change (x/y) by speed/-7.5**

The script could be done twice for both the x and y axis, or you could take the lazy route of putting them both in the same repeat (not reccomended).

**Comment by afkvido** [comment]
Lmao idk if this is scratch code or something, but all I'm seeing is this:
**if touching wall**

Alright, so how would we check _that_?

**Comment by jat354155** [comment]
It is scratch code, but it is an idea of how to implement it

**Comment by afkvido** [comment]
So... how are we going to check **if touching wall**?

**Comment by Erisfiregamer1** [comment]
@afkvido try checking it by clicking ahead of where they're going to go, prodigy wont walk the player into the wall.
if the event cancells then it's a wall and dont let them move

**Comment by afkvido** [comment]
Alright, how would you simulate a click?

**Comment by Erisfiregamer1** [comment]
Take control of mouse for a split second and click there.

**Comment by afkvido** [comment]
So we hijack the mouse, target it to the current location then the target location, and we click, if our pos doesn't change then we hit a wall?

**Comment by afkvido** [comment]
Then we make that as a completley seperate hack, cuz current WASD works by editing player location variables directly.

**Comment by afkvido** [comment]
Don't get your hopes up- I haven't actually implemented the click-to-move yet.

However, when one hack turns on, the other turns off, it's beautiful!

File: cheatGUI/src/utils/wasd.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+function ChangeX (x : number) {
+    for (let i = 0; i < 100; i++) {
+        setTimeout(() => {  _.player._playerContainer.x -= x/100; }, 50);
+    }
+}
+
+function ChangeY (y : number) {
+    for (let i = 0; i < 100; i++) {
+        setTimeout(() => {  _.player._playerContainer.y -= y/100; }, 50);
```
Comment (hostedposted) [inline cheatGUI/src/utils/wasd.ts:18]: Why setTimeout?

Comment (Erisfiregamer1) [reply cheatGUI/src/utils/wasd.ts:18]: What do you suggest we use?

Comment (afkvido) [reply cheatGUI/src/utils/wasd.ts:18]: What's wrong with setTimeout?

Comment (hostedposted) [reply cheatGUI/src/utils/wasd.ts:18]: Do you even know what it does? There is no point in doing something like that.

Comment (afkvido) [reply cheatGUI/src/utils/wasd.ts:18]: lmao yes, it runs a function after a delay,

Comment (hostedposted) [reply cheatGUI/src/utils/wasd.ts:18]: What you are doing is the same as
```ts
function ChangeY (y : number) {
    setTimeout(() => {
        for (let i = 0; i < 100; i++) {
            _.player._playerContainer.y -= y/100;
    }}, 50)
}
```

Comment (afkvido) [reply cheatGUI/src/utils/wasd.ts:18]: Tis not the same, cuz we are doing it inside the loop, not outside

Comment (Erisfiregamer1) [reply cheatGUI/src/utils/wasd.ts:18]: 🍿

Comment (afkvido) [reply cheatGUI/src/utils/wasd.ts:18]: **am i being sped**

Comment (Erisfiregamer1) [reply cheatGUI/src/utils/wasd.ts:18]: Probably, but how would I know?

Comment (afkvido) [reply cheatGUI/src/utils/wasd.ts:18]: alright we can use async/await for those functions, and put await before literally **everything** in that function, also use some form of proper wait?

**Comment by afkvido** [comment]
@Erisfiregamer1 @donaldli2020 
Take a look at... https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent

**Pull request merged**